src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Progress/CodeAnalysisProgress.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Progress;
+
+namespace Microsoft.CodeAnalysis;
+
+/// <summary>
+/// Represents the progress of an operation.  Commonly used to update a UI visible to a user when a long running
+/// operation is happening.
+/// </summary>
+internal sealed class CodeAnalysisProgress
```

Comment:
this is the core progress type we would make public.  It allows consumption code to:

1. set the description of hwat's going on to show the user.
2. add incomplete items to the overall progress (For growing a progress bar)
3. markign items as 'complete' (which would move the completed part of a progress bar forward).
4. clearing progress entirely (used for features which have a progress bar go from start to finish a few times as they do work).  

NOte: '4' is something we do in only one place (computing code actions, then applying them).  We could likely take that part out of this API and make it something that just the suggested-action-engine could do since it has access to the IUIThreadOpContext directly.